### PR TITLE
[Deploy command — 02] Add `deploy_contract` facade

### PR DESCRIPTION
**Pull request by kasperski95**
This function is based on the starkware's deploy contract. I had to rewrite it, because starkware's function was heavily integrated with argparse.

### CODE REVIEW

File: tests/integration/deploying/deploying_contract_test.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def get_available_port() -> int:
+    with Socket() as socket:
+        socket.bind(("", 0))
+        return socket.getsockname()[1]
+
+
+@pytest.fixture(name="devnet_gateway_url")
+def devnet_gateway_url_fixture():
```
Comment (MaksymilianDemitraszek) [inline tests/integration/deploying/deploying_contract_test.py:20]: I like the approach, but I think we should set scope to `session` as startup is quite expensive

File: tests/integration/deploying/deploying_contract_test.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    devnet_port = get_available_port()
```
Comment (MaksymilianDemitraszek) [inline tests/integration/deploying/deploying_contract_test.py:21]: Can we make this a fixture for consistency?

Comment (MaksymilianDemitraszek) [inline tests/integration/deploying/deploying_contract_test.py:21]: Can we make this a fixture for consistency?

File: tests/integration/deploying/deploying_contract_test.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    proc = subprocess.Popen(command)
+    time.sleep(5)
+    yield f"http://localhost:{devnet_port}"
+    proc.kill()
+
+
+@pytest.fixture(name="compiled_contract_file")
+def compiled_contract_file_fixture(tmpdir):
+    copy_file(
+        str(Path(__file__).parent / "data" / "main.json"),
```
Comment (MaksymilianDemitraszek) [inline tests/integration/deploying/deploying_contract_test.py:42]: I think we should use name `fixtures` not `data` as directory name to remain consistent.

File: protostar/commands/deploy/deploy_contract.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    code: str
+    address: str
+    transaction_hash: str
+
+
+class DeployContractException(ProtostarException):
+    pass
+
+
+async def deploy_contract(
```
Comment (MaksymilianDemitraszek) [inline protostar/commands/deploy/deploy_contract.py:28]: Can we name it `deploy` as the orginal?

File: protostar/commands/deploy/deploy_contract.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+from starkware.starknet.services.api.gateway.gateway_client import GatewayClient
+from starkware.starknet.services.api.gateway.transaction import Deploy
+from starkware.starknet.utils.api_utils import cast_to_felts
+from starkware.starkware_utils.error_handling import StarkErrorCode
+from typing_extensions import TypedDict
+
+from protostar.protostar_exception import ProtostarException
+
+
+class SuccessfulGatewayResponseFacade(TypedDict):
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/deploy/deploy_contract.py:18]: I think `SuccessfulGatewayResponse` would be enough

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
LGTM but has some issues

File: protostar/commands/deploy/deploy_contract.py (hunk 4 above)
Comment (MaksymilianDemitraszek) [reply protostar/commands/deploy/deploy_contract.py:28]: I wanted to find the original function in the starkware codebase and it mislead me

File: tests/integration/deploying/deploying_contract_test.py (hunk 3 above)
Comment (kasperski95) [reply tests/integration/deploying/deploying_contract_test.py:42]: I think we should use https://pypi.org/project/pytest-datadir/. It uses `data` directory or the name of the module.

File: protostar/commands/deploy/deploy_contract.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/deploy/deploy_contract.py:18]: I used `facade` suffix to indicate that the original dictionary may have other fields. (In the next PRs, I changed it to a data class to add a method that log response.)

[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/deploy/deploy_contract.py:18]: I still don't think `facade` is a correct word, this one just holds data and provides utilities as you mention. It doesn't simplify any logic

[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/deploy/deploy_contract.py:18]: It simplifies the interface though.

[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/deploy/deploy_contract.py]: It wraps data of three strings. we do not have`TestSuiteFacade` it's inconsistent, please rename it

**Review by MaksymilianDemitraszek — APPROVED** [review]
LGTM, but please rename the facade class, it's just wrong

File: protostar/commands/deploy/deploy_contract.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/deploy/deploy_contract.py]: We probably should rename `StarknetCompiler -> StarknetCompilerFacade` though...

[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/deploy/deploy_contract.py]: We should. `TestSuite` is our internal dataclass. I am not convinced, but I'll rename for the sake of consistency.

**Pull request merged**